### ISSUE TITLE
WASM backend

### ISSUE THREAD

**Issue by FL33TW00D** [huggingface/ratchet#146]
Whilst WebGPU support is growing, perhaps it would be prudent to have a WASM backend to ensure 100% device coverage.

**Comment by singhaki** [comment]
@FL33TW00D  have you checked this https://github.com/WasmEdge/WasmEdge ? they already have llaama.cpp, pytorch and tensorflow backend.

this project is for llm - https://github.com/LlamaEdge/LlamaEdge/tree/main

**Comment by FL33TW00D** [comment]
> @FL33TW00D have you checked this https://github.com/WasmEdge/WasmEdge ? they already have llaama.cpp, pytorch and tensorflow backend.
> 
> this project is for llm - https://github.com/LlamaEdge/LlamaEdge/tree/main

Yeah I've seen this! I think we can't really use it, but good reference!

**Comment by flatsiedatsie** [comment]
Interesting project. How is it related to Transformers.js? Since both are a project of HuggingFace, correct?

On topic: I'm working on a project that uses multiple 'engines' to achieve this.
- WebLLM when for WebGPU is available
- Wllama when it's not.

I try to choose models that run on both, and then invisibly switch engines for the user, depending on their browser.

Transformers.js runs all the smaller but vital other models, like voice recognition, TTS, OCR, embedding, etc.

**Comment by FL33TW00D** [comment]
> Interesting project. How is it related to Transformers.js? Since both are a project of HuggingFace, correct?
> 
> On topic: I'm working on a project that uses multiple 'engines' to achieve this.
> 
> * WebLLM when for WebGPU is available
> * Wllama when it's not.
> 
> I try to choose models that run on both, and then invisibly switch engines for the user, depending on their browser.
> 
> Transformers.js runs all the smaller but vital other models, like voice recognition, TTS, OCR, embedding, etc.

Ratchet is mostly experimental at the moment! Transformers.JS is production ready :)
